docs/specification/auth/oauth2.md

```diff
@@ -0,0 +1,285 @@
+    participant Client as MCP Client
+    participant AuthServer as Authorization<br> MCP Server
+    participant Resource as Resource<br>MCP Server
+
+    User->>Client: Initiate Login
+    Client->>AuthServer: Authorization Request
+    Note over Client,AuthServer: client_id, redirect_uri, scope, state
+    AuthServer->>User: Redirect to Login/Consent/Configuration
+    User->>AuthServer: Authenticate & Authorize
+    AuthServer->>Client: Authorization Code
+    Note over AuthServer,Client: Via redirect_uri with state
+    Client->>AuthServer: Token Request
+    Note over Client,AuthServer: auth code, client_id, client_secret
+    AuthServer->>Client: Access & Refresh Tokens
+    Client->>Resource: Access prompts/resources/tools
+    Resource->>Client: Access roots/sampling
+    
+    Note right of Client: When token expires...
+    Client->>AuthServer: Refresh Token Request
+    AuthServer->>Client: New Access Token
+```
+
+## Capabilities
+
+Clients supporting OAuth 2.0 **MUST** declare it during initialization:
+
+```json
+{
+  "capabilities": {
+    "auth": {
+      "oauth2": true
+    }
+  }
+}
+```
+
+Servers supporting OAuth 2.0 **MUST** include their capabilities:
+
+```json
+{
+  "capabilities": {
+    "auth": {
+      "oauth2": {
+        "token": true,
+        "revoke": true
+      }
+    }
+  }
+}
+```
+
+## Authentication Flows
+
+### Get Access Token Flow
+
+1. Initial Request
+
+When a client without valid credentials attempts to initialize, the server **MUST** respond with an OAuth 2.0 authorization requirement in the form of an error:
+
+```json
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "error": {
+    "code": -32001,
+    "message": "OAuth 2.0 authorization required",
+    "data": {
+      "authRequest": {
+        "authorizationUrl": "<scheme>://<domain>/oauth2",
+        "clientId": "<client_id>",
+        "scope": ["read", "write"],
+        "responseType": "code",
+        "state": "state"
+      }
+    }
+  }
+}
+```
+The authorization URL **SHOULD** prompt the user to configure any credentials as needed.
+
+2. Authorization Code Exchange
+
+After receiving the authorization code, the client **MUST** exchange it for tokens:
+
+```json
+{
+  "jsonrpc": "2.0",
+  "id": 2,
+  "method": "auth/oauth2/token",
+  "params": {
+    "grant_type": "authorization_code",
+    "code": "auth_code",
+    "state": "xyz"
+  }
+}
+```
+
+The server responds with tokens:
+
+```json
+{
+  "jsonrpc": "2.0",
+  "id": 2,
+  "result": {
+    "access_token": "access_token",
+    "refresh_token": "refresh_token",
+    "expires_in": 3600
+  }
+}
+```
+
+### Valid Access Token Flow
+
+Once a client has obtained an access token, it **MUST** include it in the initialization request:
+
+```json
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "method": "initialize",
+  "params": {
+    "protocolVersion": "...",
+    "capabilities": {
+      "auth": {
+        "oauth2": true
+      }
+    },
+    "headers": {
+      "Authorization": "Bearer access_token"
+    }
```

Comment:
The format of this is a bit weird to me, and I think we need to support multiple auth schemes here too, so something like:
```suggestion
    "auth": {
      "oauth2": {
        "access_token": "..."
      }
    }
```